services/venice-controller/src/main/java/com/linkedin/venice/controller/ControllerInstanceTagRefresher.java

```
@@ -0,0 +1,82 @@
+package com.linkedin.venice.controller;
+
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.helix.SafeHelixManager;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.helix.ConfigAccessor;
+import org.apache.helix.PreConnectCallback;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * PreConnectCallback is called after the participant registers its info in the cluster but before marking itself as
+ * LIVEINSTANCE. If the instance is already a participant of the cluster, then we need this logic to update Instance tags
+ * in the case of KSAP assigning this controller participant to another cluster
+ */
+public class ControllerInstanceTagRefresher implements PreConnectCallback {
+  private static final Logger LOGGER = LogManager.getLogger(ControllerInstanceTagRefresher.class);
+  private SafeHelixManager helixManager;
+  private final VeniceControllerMultiClusterConfig multiClusterConfigs;
+
+  public ControllerInstanceTagRefresher(
+      SafeHelixManager helixManager,
+      VeniceControllerMultiClusterConfig multiClusterConfigs) {
+    this.helixManager = helixManager;
+    this.multiClusterConfigs = multiClusterConfigs;
+  }
+
+  @Override
+  public void onPreConnect() {
+    try {
+      String instanceName = helixManager.getInstanceName();
+      ConfigAccessor configAccessor = helixManager.getConfigAccessor();
+      InstanceConfig instanceConfig = configAccessor.getInstanceConfig(helixManager.getClusterName(), instanceName);
+
+      if (instanceConfig == null) {
+        LOGGER.warn("No InstanceConfig found for {}. Creating new config.", instanceName);
+        instanceConfig = new InstanceConfig(instanceName);
```

We add the port in `connectToControllerCluster`. Can we do the same here?

```
InstanceConfig.Builder defaultInstanceConfigBuilder =
        new InstanceConfig.Builder().setPort(Integer.toString(multiClusterConfigs.getAdminPort()));
```

---

I had considered this, I wanted to narrow the scope of `ControllerInstanceTagRefresher`, especially since the port remains the same across different clusters

---

Since we're creating a new `InstanceConfig`, it won't have the port that we set in `connectToControllerCluster`. This may cause issues if the default port in `InstanceConfig` isn't what we're expecting.

---

I think we should throw an exception here. It should be expected that the instance is created before this function is invoked

---

Ok, throwing an exception if the instance config is expected to exist